### PR TITLE
fix(mssql): support top syntax for limiting queries

### LINKED ISSUE #8074: Common Table Expressions (CTE's) do not work for Microsoft SQL Server database
### Problem
Common Table Expressions (CTE) queries are not working as expected in SQL Lab against SQL Server databases. These are queries written using the `WITH` clause.

### Steps to Reproduce
Start superset using:

```bash
superset run -p 8080 --with-threads --reload --debugger
```
Add a SQL Server database and tables in `Sources`.

Make sure to place a check for `Allow DML` and ensure that it is selected and set to True.

Navigate to SQL Lab. Run the following query:

```sql
WITH t AS (SELECT 1 a) SELECT * FROM t
```

#### Expected Result

```text
------
| a  |
------
| 1  |
------
```

#### Observed Result

The following error message is seen:

```text
2019-08-20 14:07:53,005:ERROR:root:('42000', "[42000] [FreeTDS][SQL Server]Incorrect syntax near the keyword 'WITH'. (156) (SQLExecDirectW)")
Traceback (most recent call last):
  File "/home/user/anaconda3/envs/superset/lib/python3.6/site-packages/superset/sql_lab.py", line 190, in execute_sql_statement
    db_engine_spec.execute(cursor, sql, async_=True)
  File "/home/user/anaconda3/envs/superset/lib/python3.6/site-packages/superset/db_engine_specs.py", line 421, in execute
    cursor.execute(query)
pyodbc.ProgrammingError: ('42000', "[42000] [FreeTDS][SQL Server]Incorrect syntax near the keyword 'WITH'. (156) (SQLExecDirectW)")
2019-08-20 14:07:53,017:INFO:werkzeug:127.0.0.1 - - [20/Aug/2019 14:07:53] "POST /superset/sql_json/ HTTP/1.1" 500 -
```

### System and Environment Details
Database: `Microsoft SQL Server 2017`

Operating System: `Red Hat Enterprise Linux 7 x86_64`
Python: `Python 3.6.7` (Anaconda Python)

Superset:
```text
Superset 0.33.0rc1
flask 1.1.1
werkzeug 0.15.5
```
Other packages:
```text
sqlalchemy 1.3.7
sqlparse 0.3.0
unixodbc 2.3.7
pyodbc 4.0.23
```

```text
FreeTDS 1.1.11
```
#### Additional Notes
The query runs as expected against the same database:
- Directly in `Microsoft SQL Server Management Studio 17`
- Through FreeTDS (by running `tsql -S <FQDN>` and then executing the query)
- Through pyODBC and SQLAlchemy in a python program.

The query runs as expected against a PostGreSQL database through SQL Lab.

### PR DESCRIPTION
**Pull request by sujiplr**
MSSQL is not supporting LIMIT syntax in SQLs. For limiting the rows, MSSQL having a different keyword TOP. Added fixes for handling the TOP and LIMIT clauses based on the database engines. Made this fix as generalised one for handling different database engines.

### SUMMARY
This fix will help to start supporting TOP clause in SQLs. And also along with CTE queries.

### BEFORE SCREENSHOTS

Pre-error-with-cte

### AFTER SCREENSHOTS

Post-fix-working with top and cte

### TESTING INSTRUCTIONS
You can test the SQLs as per the above screenshots. 

### ADDITIONAL INFORMATION

- [x] Has associated issue: closes #8074
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/db_engine_specs/base.py [hunk 1]
```diff
@@ -649,6 +653,57 @@ def apply_limit_to_sql(
...
+        Alters the SQL statement to apply a TOP clause
+        :param _limit: Maximum number of rows to be returned by the query
+        :param sql: SQL query
+        :param database: Database instance
+        :return: SQL query with top clause
+        """
+        sql_statement = sqlparse.format(sql, strip_comments=True)
+        query_limit: Optional[int] = sql_parse._extract_top_from_query(sql_statement)
+        td_sel_keywords = {"SELECT","SEL"}
+        td_top_keywords = {"TOP","SAMPLE"}
```
Comment (villebro) [inline superset/db_engine_specs/base.py:670]: `SEL` and `SAMPLE` seem to be specific to Teradata. We should probably move these into `BaseEngineSpec` like this:

```python
class BaseEngineSpec:
    ...
    # all keywords that should be treated as SELECTs
    select_keywords = {"SELECT"}
    # For engines that use TOP instead of LIMIT keyword, 
    list all keywords that should be treated similar top TOP keywords
    top_keywords = {"TOP"}
```

Then for Teradata these could be set to
```python
    select_keywords = {"SELECT","SEL"}
    top_keywords = {"TOP","SAMPLE"}
```

and for MSSQL it would probably be ok to go with the defaults (I assume `SEL` and `SAMPLE` aren't valid for MSSQL).

File: superset/sql_parse.py [hunk 2]
```diff
@@ -77,6 +77,29 @@ def _extract_limit_from_query(statement: TokenList) -> Optional[int]:
...
+    token = [part for part in token if part]
+    top = None
+    for i, _ in enumerate(token):
+        if token[i].upper() in td_top_keyword and len(token) - 1 > i:
+            try:
+                top = int(token[i + 1])
+            except ValueError:
+                top = None
+            break
+    return top
```
Comment (villebro) [inline superset/sql_parse.py:101]: Here we'd need to pass in the `top_keywords` from the db engine spec to make it possible to customize supported keywords per engine. Something like this:
```suggestion
def _extract_top_from_query(statement: TokenList, top_keywords: Set[str]) -> Optional[int]:
    """
    Extract top clause value from SQL statement.

    :param statement: SQL statement
    :return: top value extracted from query, None if no top value present in statement
    """

    str_statement = str(statement)
    str_statement = str_statement.replace("\n", " ").replace("\r", "")
    token = str_statement.rstrip().split(" ")
    token = [part for part in token if part]
    top = None
    for i, _ in enumerate(token):
        if token[i].upper() in top_keywords and len(token) - 1 > i:
            try:
                top = int(token[i + 1])
            except ValueError:
                top = None
            break
    return top
```

**Review by villebro — COMMENTED** [review]
First pass comments. Let's also ping the original authors of the Teradata TOP functionality when this is ready 👍

File: superset/db_engine_specs/base.py (hunk 1 above)
Comment (sujiplr) [reply superset/db_engine_specs/base.py:670]: Made this change

File: superset/db_engine_specs/base.py [hunk 3]
```diff
@@ -638,8 +647,8 @@ def apply_limit_to_sql(
             sql = sql.strip("\t\n ;")
             qry = (
                 select("*")
-                .select_from(TextAsFrom(text(sql), ["*"]).alias("inner_qry"))
-                .limit(limit)
+                    .select_from(TextAsFrom(text(sql), ["*"]).alias("inner_qry"))
+                    .limit(limit)
```
Comment (villebro) [inline superset/db_engine_specs/base.py:651]: I expect these changes to cause linting errors

File: superset/db_engine_specs/base.py [hunk 4]
```diff
@@ -697,6 +770,7 @@ def get_cte_query(cls, sql: str) -> Optional[str]:
 
         return None
 
+
```
Comment (villebro) [inline superset/db_engine_specs/base.py:773]: nit: This will also cause a linting error
```suggestion
```

File: superset/db_engine_specs/mssql.py [hunk 5]
```diff
@@ -48,6 +48,9 @@ class MssqlEngineSpec(BaseEngineSpec):
     limit_method = LimitMethod.WRAP_SQL
     max_column_name_length = 128
     allows_cte_in_subquery = False
+    allow_limit_clause = False
+    sel_keywords = {"SELECT"}
+    top_keywords = {"TOP"}
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py:53]: wasn't this called `select_keywords` in `BaseEngineSpec`? Also, no need to override if the property is unchanged in the extending class.

**Review by villebro — COMMENTED** [review]
Second pass comments

File: superset/db_engine_specs/mssql.py (hunk 5 above)
Comment (sujiplr) [reply superset/db_engine_specs/mssql.py:53]: Corrected this.

File: superset/db_engine_specs/base.py (hunk 3 above)
Comment (sujiplr) [reply superset/db_engine_specs/base.py:651]: Corrected this

File: superset/db_engine_specs/base.py [hunk 6]
```diff
@@ -300,6 +299,16 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
...
+    # Whether allow LIMIT clause in the SQL
+    # If True, then the database engine is allowed for LIMIT clause
+    # If False, then the database engine is allowed for TOP clause
+    allow_limit_clause = True
+    # This list will give the default list of keywords for select statements
+    # to consider in the limit/top SQl parsing
+    select_keywords = {"SELECT"}
+    # This list will give the default list of keywords for data limit statements
+    # to consider in the limit/top SQl parsing
+    top_keywords = {"LIMIT"}
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:311]: corrected this

File: superset/db_engine_specs/base.py [hunk 7]
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
...
+                    next_is_limit_token = False
+            new_tokens.append(token)
+        sql = " ".join(new_tokens)
+        return cte + sql
+
+    def top_not_in_sql(sql: str, top_words: Set[str]) -> bool:
+        for top_word in top_words:
+            if top_word.upper() in sql.upper():
+                return False
+        return True
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:723]: Corrected this

File: superset/db_engine_specs/base.py [hunk 8]
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
...
+            str_statement = str(sql_remainder)
+            cte = cte + "\n"
+        else:
+            cte = ""
+            str_statement = str(sql)
+        str_statement = str_statement.replace("\n", " ").replace("\r", "")
+
+        tokens = str_statement.rstrip().split(" ")
+        tokens = [token for token in tokens if token]
+        if cls.top_not_in_sql(str_statement, td_top_keywords):
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:698]: Corrected this

File: superset/db_engine_specs/base.py [hunk 9]
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
...
+        :return: SQL query with top clause
+        """
+
+        td_sel_keywords = cls.select_keywords
+        td_top_keywords = cls.top_keywords
+        cte = None
+        sql_remainder = None
+        sql_statement = sqlparse.format(sql, strip_comments=True)
+        query_limit: Optional[int] = sql_parse._extract_top_from_query(sql_statement,
+                                                                       td_top_keywords)
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:679]: Corrected this

File: superset/db_engine_specs/base.py [hunk 10]
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
...
+        """
+        Alters the SQL statement to apply a TOP clause
+        :param _limit: Maximum number of rows to be returned by the query
+        :param sql: SQL query
+        :param database: Database instance
+        :return: SQL query with top clause
+        """
+
+        td_sel_keywords = cls.select_keywords
+        td_top_keywords = cls.top_keywords
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:674]: Corrected this, instead of assignment referenced as class property value.

File: superset/db_engine_specs/base.py (hunk 4 above)
Comment (sujiplr) [reply superset/db_engine_specs/base.py:773]: Corrected.

**Review by villebro — COMMENTED** [review]
Some last nit comments, but I think this is pretty much ready for a final test pass + a few added unit tests to the MSSQL test suite. Also, I suggest checking failing workflows on CI, as there appears to be simple linting issues that need to be addressed.

File: superset/db_engine_specs/base.py [hunk 11]
```diff
@@ -80,7 +80,6 @@
 
 logger = logging.getLogger()
 
-
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:83]: Corrected.

File: superset/db_engine_specs/base.py [hunk 12]
```diff
@@ -300,6 +299,16 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
...
+    # Whether allow LIMIT clause in the SQL
+    # If True, then the database engine is allowed for LIMIT clause
+    # If False, then the database engine is allowed for TOP clause
+    allow_limit_clause = True
+    # This list will give keywords for select statements
+    # to consider for the engines with TOP SQl parsing
+    select_keywords: Set[str] = None
+    # This list will give the keywords for data limit statements
+    # to consider for the engines with TOP SQl parsing
+    top_keywords: Set[str] = None
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:311]: Corrected

File: superset/db_engine_specs/base.py [hunk 13]
```diff
@@ -649,6 +658,69 @@ def apply_limit_to_sql(
...
+        query_limit: Optional[int] = sql_parse._extract_top_from_query(sql_statement,
+                                                                       cls.top_keywords)
+        if not _limit:
+            final_limit = query_limit
+        elif int(query_limit or 0) < _limit and query_limit is not None:
+            final_limit = query_limit
+        else:
+            final_limit = _limit
+        if not cls.allows_cte_in_subquery:
+            cte, sql_remainder = sql_parse.get_cte_reminder_query(sql_statement)
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/db_engine_specs/base.py:685]: Corrected.

File: superset/sql_parse.py [hunk 14]
```diff
@@ -78,6 +78,56 @@ def _extract_limit_from_query(statement: TokenList) -> Optional[int]:
...
+        if token[i].upper() in top_keyword and len(token) - 1 > i:
+            try:
+                top = int(token[i + 1])
+            except ValueError:
+                top = None
+            break
+    return top
+
+
+def get_cte_reminder_query(sql: str) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/sql_parse.py:105]: Corrected

File: superset/sql_lab.py [hunk 15]
```diff
@@ -287,15 +287,27 @@ def execute_sql_statement(  # pylint: disable=too-many-arguments,too-many-locals
...
 
 
+def apply_cte(
+    database: Database, sql: str
+) -> str:
+    # Use the Basengine function to parse the SQL for CTE
+    sql = database.db_engine_spec.get_cte_query(sql)
+    return sql
+
+
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/sql_lab.py:297]: removed this function.

File: superset/sql_parse.py [hunk 16]
```diff
@@ -30,7 +30,8 @@
     Token,
     TokenList,
 )
-from sqlparse.tokens import DDL, DML, Keyword, Name, Punctuation, String, Whitespace
+from sqlparse.tokens import DDL, DML, Keyword, Name, Punctuation, String, Whitespace, \
+    CTE
```
[the remark this answers is not in the document]
Comment (sujiplr) [reply superset/sql_parse.py:34]: Corrected.

File: superset/sql_lab.py [hunk 17]
```diff
@@ -292,10 +292,14 @@ def apply_limit_if_exists(
...
-        # to test whether there are more rows than the limit.
+        # to test whether there are more rows than the limit. According to the DB
+        # Engine support it will choose top or limit parse
         # Later, the extra row will be dropped before sending
         # the results back to the user.
-        sql = database.apply_limit_to_sql(sql, increased_limit, force=True)
+        if database.db_engine_spec.allow_limit_clause:
+            sql = database.apply_limit_to_sql(sql, increased_limit, force=True)
+        else:
+            sql = database.apply_top_to_sql(sql, increased_limit)
```
Comment (villebro) [inline superset/sql_lab.py:302]: nit: this check should probably be done in `Database. apply_top_to_sql` so callers don't need to be aware of LIMIT/TOP.

File: superset/sql_parse.py [hunk 18]
```diff
@@ -41,7 +50,6 @@
 CTE_PREFIX = "CTE__"
 logger = logging.getLogger(__name__)
 
-
```
Comment (villebro) [inline superset/sql_parse.py:44]: nit: maybe add this back to avoid making unrelated changes

**Review by villebro — COMMENTED** [review]
Two minor nits

File: superset/sql_parse.py (hunk 18 above)
Comment (sujiplr) [reply superset/sql_parse.py:44]: Done.

File: superset/sql_lab.py (hunk 17 above)
Comment (sujiplr) [reply superset/sql_lab.py]: done

File: superset/models/core.py [hunk 19]
```diff
@@ -488,7 +488,13 @@ def select_star(  # pylint: disable=too-many-arguments
     def apply_limit_to_sql(
         self, sql: str, limit: int = 1000, force: bool = False
     ) -> str:
-        return self.db_engine_spec.apply_limit_to_sql(sql, limit, self, force=force)
+        if self.db_engine_spec.allow_limit_clause:
+            return self.db_engine_spec.apply_limit_to_sql(sql, limit, self, force=force)
+        else:
+            return self.db_engine_spec.apply_top_to_sql(sql, limit)
+
+    def apply_top_to_sql(self, sql: str, limit: int = 1000, force: bool = False) -> str:
```
Comment (villebro) [inline superset/models/core.py:496]: I don't think this is needed anymore

File: superset/sql_lab.py [hunk 20]
```diff
@@ -290,9 +290,12 @@ def execute_sql_statement(  # pylint: disable=too-many-arguments,too-many-locals
 def apply_limit_if_exists(
     database: Database, increased_limit: Optional[int], query: Query, sql: str
 ) -> str:
+    print("apply_limit_if_exists")
+    print(database.db_engine_spec.allow_limit_clause)
```
Comment (villebro) [inline superset/sql_lab.py:294]: please remove these

File: superset/sql_parse.py (hunk 18 above)
Comment (villebro) [reply superset/sql_parse.py:44]: removed line seems to still be here

File: superset/db_engine_specs/base.py (hunk 11 above)
Comment (villebro) [inline superset/db_engine_specs/base.py:83]: removed newline seems to still be here

File: superset/sql_lab.py (hunk 20 above)
Comment (sujiplr) [reply superset/sql_lab.py:294]: Done

File: superset/models/core.py (hunk 19 above)
Comment (sujiplr) [reply superset/models/core.py:496]: Done

File: superset/db_engine_specs/base.py (hunk 11 above)
Comment (sujiplr) [reply superset/db_engine_specs/base.py:83]: done

File: superset/db_engine_specs/mssql.py [hunk 21]
```diff
@@ -48,6 +48,9 @@ class MssqlEngineSpec(BaseEngineSpec):
     limit_method = LimitMethod.WRAP_SQL
     max_column_name_length = 128
     allows_cte_in_subquery = False
+    allow_limit_clause = False
+    select_keywords = {"SELECT"}
+    top_keywords = {"TOP"}
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py:53]: these can be removed (they're identical to the defaults in `BaseEngineSpec`)

**Review by villebro — COMMENTED** [review]
Last nit, after this I believe this is good to go

**Review by villebro — APPROVED** [review]
LGTM! Thanks for all the hard work here; this is a major step forward in ensuring Superset works properly on MSSQL, Synapse and other Azure db products! ❤️

**Comment by sfirke** [comment]
I am noodling on https://github.com/apache/superset/issues/21027#issuecomment-1516398570 and found this PR via blame view on the relevant code.  @sujiplr do you have thoughts on that issue - modifying the codebase to support use of `DISTINCT` in SQL Lab for SQL Server?  The current code injects `TOP 1000` in the wrong place for MSSQL so it's `SELECT TOP 1000 DISTINCT` when it needs to be `SELECT DISTINCT TOP 1000`.
